code/move/solution/trymove.sol.cpp

```diff
@@ -1,39 +1,87 @@
-#include <iostream>
-#include <vector>
+#include <memory>
 #include <random>
-#include "NVector.sol.hpp"
-
-template<typename T>
-void swap(std::vector<T> &v, int a, int b) {
-    T mem = std::move(v[a]);
-    v[a] = std::move(v[b]);
-    v[b] = std::move(mem);
-    // even better
-    //swap(v[a], v[b]);
-}
+#include <vector>
 
-template<typename T>
-void randomize(std::vector<T> &v) {
-    // we randomize via len random inversions
-    std::size_t len = v.size();
-    std::default_random_engine e;
-    std::uniform_int_distribution<std::size_t> d{0, len};
-    for (std::size_t i = 0; i < len; i++) {
-        int a = d(e);
-        int b = d(e);
-        swap(v, a, b);
+/**
+ * An attempt to make a movable class. It allocates memory on construction, and has
+ * array semantics. (Prefer std::array or std::vector instead of hand-crafted solutions, though.)
+ * We made it copyable by providing copy constructor and assignment operator, but note that we
+ * are violating the rule of 5, so it's not moveable yet.
+*/
+class Moveable {
+    static constexpr std::size_t size = 10000;
+
+public:
+    Moveable() = default;
+    Moveable(Moveable const & other) :
+        Moveable()
+    {
+        // Delegate copying to the assignment operator
+        *this = other;
     }
-}
+    Moveable & operator=(const Moveable & other)
+    {
+        if (other.m_storage != m_storage) {
```

Comment:
This is protection against self-assignment, is it? Then why are you not using the canonical form?
```suggestion
        if (&other != this) {
```

Reply:
I'm not aware of a canonical form.

Reply:
See e.g. here: https://en.cppreference.com/w/cpp/language/operators#Assignment_operator. You can guard against self-assignment by comparing `other`'s identity with `this`.